### PR TITLE
Avoid that cor(x,x) != 1...

### LINKED ISSUE #17420: Cor sometimes returns values > 1
```
julia> cor(1:100, 1:100) > 1
true
```
This happens due to FP precision, but it can potentially break downstream functions that rely on -1 <= cor <= 1

### PR DESCRIPTION
**Pull request by andreasnoack**
...by trading a `sqrt` for two `/` and use `clamp` to avoid that `cor(x,y) > 1`.

Make sure that the loop vectorizes

Convert `corzm` into `corm` to avoid allocation of tempraries.

Fixes #17420

### CODE REVIEW

File: base/statistics.jl [hunk 1]
```diff
@@ -368,25 +368,6 @@ function corzm(x::AbstractMatrix, vardim::Int=1)
     c = unscaled_covzm(x, vardim)
     return cov2cor!(c, sqrt!(diag(c)))
 end
-function corzm(x::AbstractVector, y::AbstractVector)
```
Comment (tkelman) [inline base/statistics.jl]: are packages calling this method? may be safer to deprecate

Comment (andreasnoack) [inline base/statistics.jl]: I don't think so. I checked and I've only seen @simonster using it for WIP. It's not exported so deprecating it also a little bit messy.

**Comment by andreasnoack** [comment]
@tkelman Can this fix be merged?

**Comment by tkelman** [comment]
if it's non-breaking and has been reviewed by someone like @simonbyrne or equivalent, I guess

**Comment by StefanKarpinski** [comment]
Yeah, this strikes me as a bug fix, so eligible for 0.5 before or after RC.

**Comment by simonbyrne** [comment]
LGTM.

**Pull request merged**